fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+        return result
+
+    def drop_stale_feature_table(self, table_name):
+        """Try to remove a table even if views depend on it
+
+        Args:
+            table_name [string]: Name of table to be dropped
+        """
+
+        query = "DROP TABLE IF EXISTS {} CASCADE;".format(table_name)
+        self.execute_query_from_string(query)
+        return None
+
+    def generate_table_undefended_frontpage_links(self):
+        """This method creates a table of exampleids that were
+        present at the time of feature generation that the unified
+        view can join on. The table contains one integer column, exampleid.
+        """
+
+        self.drop_stale_feature_table("undefended_frontpage_examples")
+
+        query = ("CREATE TABLE features.undefended_frontpage_examples AS ( "
+                 "SELECT foo.exampleid FROM ( SELECT exampleid, "
+                 "count(*) FROM raw.frontpage_traces GROUP BY exampleid) foo);")
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_temp_packet_positions(self, outgoing_only=True):
+        """This method takes all rows in raw.frontpage_traces
+        and creates a temporary table packet_positions
+        with the following format:
+
+        exampleid  | rank     |  ingoing
+        (integer)  | (bigint) |  (boolean)
+        ------------------------------------
+        9          | 4        | f
+        9          | 5        | f
+        9          | 15       | f
+
+        where rank is the "position" in the trace for all
+        cells. For example, for the first outgoing cell in the
+        trace with exampleid=9 was the 4th packet in the trace.
+        This table is used by the position-based feature
+        generation functions.
+
+        Args:
+            outgoing_only [boolean]: describes whether or not we should
+                only use outgoing cells to create this table
+        """
+
+        self.drop_stale_feature_table("packet_positions")
+
+        if outgoing_only:
+            where_only_outgoing = "WHERE ingoing <= false"
+        else:
+            where_only_outgoing = ""
+
+        query = ("CREATE TEMP TABLE packet_positions AS    "
+                 "(SELECT exampleid, rank, ingoing                  "
+                 "  FROM (                                          "
+                 "    SELECT                                        "
+                 "      ROW_NUMBER() OVER                           "
+                 "      (PARTITION BY exampleid ORDER BY t_trace)   "
+                 "      AS rank,                                    "
+                 "      t.*                                         "
+                 "    FROM raw.frontpage_traces t)                  "
+                 "x {} );").format(where_only_outgoing)
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_table_outgoing_cell_positions(self, num_cells=500):
```

Comment:
I do see where you’re going here, however for context: there are some queries that use both of these functions and then others that use only one (e.g. generate_table_binned_counts only uses the first method: `_create_temp_packet_positions`). Also notice that the second function `_create_table_outgoing_cell_positions` expects the first temporary table to already exist, so it’s not trivial (and is quite awkward) to combine the two in a generic way… I recommend we table for now